File: web/status-history/src/test/java/net/twisterrob/travel/statushistory/HtmlValidator.java
package net.twisterrob.travel.statushistory;

import java.io.StringReader;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.List;

import org.w3c.tidy.Tidy;
import org.w3c.tidy.TidyMessage;

import static org.junit.Assert.assertEquals;

public class HtmlValidator {

	public static void assertValidHtml(String html) {
		Tidy tidy = new Tidy();
		tidy.setQuiet(true);
		tidy.setShowWarnings(true);

		List<TidyMessage> messages = new ArrayList<>();
		tidy.setMessageListener(messages::add);
		tidy.parse(new StringReader(html), new StringWriter());
		assertEquals(html + "\n" + messages, 0, messages.size());
	}
}

File: web/status-history/src/test/java/net/twisterrob/travel/statushistory/test/HtmlValidator.kt
package net.twisterrob.travel.statushistory.test

import org.junit.jupiter.api.Assertions.assertEquals
import org.w3c.tidy.Tidy
import org.w3c.tidy.TidyMessage
import java.io.Writer

object HtmlValidator {

	fun assertValidHtml(html: String) {
		val tidy = Tidy()
		tidy.quiet = true
		tidy.showWarnings = true

		val messages: MutableList<TidyMessage> = mutableListOf()
		tidy.setMessageListener(messages::add)
		tidy.parse(html.reader(), null as Writer?)
		assertEquals(0, messages.size, "${html}\n${messages}")
	}
}
